Add the parsers, indexer, and downloader

noahfalk opened the pull request:
This has initial implementations of a lot of the functionality we need, but the projects still need some massaging to be a good citizens in the repo.

noahfalk commented:
@tmat - ptal

tmat commented on src/FileFormats.ELF/ELFFile.cs:
```diff
@@ -0,0 +1,288 @@
+
+    public class ELFFile
+    {
+        IAddressSpace _dataSource;
```
```private readonly``` on all these fields?

tmat commented on src/FileFormats.ELF/ELFFile.cs:
```diff
@@ -0,0 +1,288 @@
+            return segments;
+        }
+
+        Reader CreateVirtualAddressReader()
```
```private```

tmat commented on src/FileFormats.ELF/ELFFile.cs:
```diff
@@ -0,0 +1,288 @@
+        }
+
+        #region Validation Rules
+        public ValidationRule IsHeaderPhOffValid
```
What is "PhOff"?

tmat commented:
We need to convert .xproj to .csproj files. I'll take care of that, for now let's just avoid adding these xproj files into the .sln so that they don't get built.

noahfalk commented:
@tmat - ptal
I needed to update the xproj to csproj in order to run the codeformatter tool. I had to edit the projects by hand because VS didn't want to add references, but it seems to read and build them OK. I think this addresses all the feedback + all the todo items that would be useful before checkin.

tmat commented:
Editing by hand is currently necessary, until we get a better project system in VS.

tmat commented on src/FileFormats/FileFormats.csproj:
```diff
@@ -0,0 +1,33 @@
+    <NuGetTargetMoniker>.NETStandard,Version=v1.5</NuGetTargetMoniker>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+    <NoWarn>1570;1591</NoWarn>
```
Probably want these in Release build as well.

tmat commented:
:+1:
